static/app/views/dashboardsV2/layoutUtils.tsx

```diff
@@ -188,3 +189,30 @@ export function assignDefaultLayout(
   });
   return newWidgets;
 }
+
+export function enforceWidgetHeightValues(widget: Widget): Widget {
+  const {displayType, layout} = widget;
+  const nextWidget = {
+    ...widget,
+  };
+  const minH = getDefaultWidgetHeight(displayType);
+  const nextLayout = {
+    ...layout,
+    h: Math.max(layout?.h ?? minH, minH),
+    minH,
+  };
+  return {...nextWidget, layout: nextLayout};
+}
+
+export function generateWidgetsAfterCompaction(widgets: Widget[]) {
```

Comment:
Thinking of also applying this after duplication, but I'll do that in a separate PR